examples/bond/assets/bond-dapp-stateful.py

```diff
@@ -0,0 +1,255 @@
+        App.globalPut(maturity_date, Btoi(Txn.application_args[3])),
+        App.globalPut(coupon_value, Btoi(Txn.application_args[4])),
+        App.globalPut(epoch, Btoi(Txn.application_args[5])),
+        App.globalPut(current_bond, Btoi(Txn.application_args[6])),
+        App.globalPut(max_amount, Btoi(Txn.application_args[7])),
+        App.globalPut(bond_token_creator, Txn.application_args[8]),
+        Return(Int(1))
+    ])
+
+    # Update issuer address in this contract
+    # only store manager can update issuer
+    # Expected arguments: [Bytes("update_issuer_address"), issuer_address]
+    update_issuer = Seq([
+        # asserts if sender is store manager
+        Assert(
+             And(
+                Txn.sender() == App.globalGet(store_manager),
+                basic_checks
+            )
+        ),
+        App.globalPut(issuer_address, Txn.application_args[1]),
+        Return(Int(1))
+    ])
+
+    # Update issue price
+    # only bond app manager can update issue price
+    # expected arguments: [Bytes("update_issue_price"), new_issue_price]
+    on_update_issue_price = Seq([
+        # asserts if sender is store manager
+        Assert(
+             And(
+                Txn.sender() == App.globalGet(store_manager),
+                basic_checks
+            )
+        ),
+        App.globalPut(issue_price, Txn.application_args[1]),
+        Return(Int(1))
+    ])
+
+    # Issue bond tokens to issuer
+    # Expected arguments: [Bytes("issue")]
+    on_issue = Seq([
+        # assert sender is bond token creator and receiver is issuer address
+        Assert(
+            And(
+                Gtxn[0].type_enum() == TxnType.AssetTransfer,
+                Gtxn[0].asset_sender() == App.globalGet(bond_token_creator),
+                Gtxn[0].asset_receiver() == App.globalGet(issuer_address),
+                # verify current bond is being transferred
+                Gtxn[0].xfer_asset() == App.globalGet(current_bond),
+                basic_checks
+            )
+        ),
+        Return(Int(1))
+    ])
+
+    # Buy transaction, sent from buyer
+    # Expected arguments: [Bytes("buy")]
+    # 1. amount×BondApp.issue_price + algo_tx_price
+    # - ALGO payment from buyer to the issuer. algo_tx_price is required to cover the TX2 cost.
+    # 2. B_i ASA transfer from issuer to the buyer.
+    on_buy = Seq([
+        Assert(
+            And(
+                basic_checks,
+                # verify payment
+                Gtxn[0].type_enum() == TxnType.Payment,
+                # verify buying amount,
+                Gtxn[0].amount() >= Add(Mul(Gtxn[1].asset_amount(), App.globalGet(issue_price)), Gtxn[1].fee()),
+                # verify ASA transfer
+                Gtxn[1].type_enum() == TxnType.AssetTransfer,
+                # verify current bond is being transferred
+                Gtxn[1].xfer_asset() == App.globalGet(current_bond),
+                Gtxn[1].asset_receiver() == Gtxn[0].sender()
```

Comment:
there are three transactions in buy operation:
1. algo transfer from buyer to issuer (fees paid by buyer because buyer is transferring algo)
    - this amount includes fee of tx2 + amount required to buy tokens
2. tokens transfer from issuer to buyer(`fee deducted from amount issuer got which already has fees`)
3. call to bond-dapp (by buyer)
`Buyer pays all the fees in above tranactions.`

Reply:
> this amount includes fee of tx2 + amount required to buy tokens

ok, so this is checked at line 124 right? Could you please update the comment, that we're verifying fee of `tx1` is being checked as well, that it's paid by `tx0`.

BDW, @amityadav0  @robert-zaremba  we also have pooled transaction fees support now, we can utilize that as well. Buyer don't need to explicitly send amount of fees in algos to the issuer. So in this case for example:
```py
fees(tx0 OR tx2) = 2000 (to cover fee of tx1 as well)
fee(tx1) = 0 // issuer doesn't need to pay fee (during token transfer)
```
And in contract we can just assert `Gtxn[1].fee == 0` (verify issuer doesn't need to pay the fee). Just a thought, current implementation works as well.

Reply:
good idea. yes, i have already updated them and added checks that issuer doesn't pay fee, Gtxn[1].fee == Int(0).

Reply:
pooled fees is better also, because attacker can easily bypass our scripts, send only the amount required this will result in deduction in fee from issuer.

Reply:
Hmm, I'm not sure if pooled fee will work there. If user will not attach more fee, then the fee will be charged from the lsig account for the transaction where we send tokens from issuer to the user.
We need to investigate it and add appropriate comments in the code and scripts.

Reply:
I have added checks for that `Gtxn[1].fee == Int(0)` for issuer, dexLsig transactions, so it is not possible for user to not attach less fees, If user does that the transaction will always fail. we ensure this in stateful contract that issuer, dex don't pay fees when a user interacts with contract. Adding appropriate comments.

Reply:
OK, `Gtxn[1].fee == Int(0)` is a good solution. Please review other places where where a user is interacting with the contract - to make sure he will pay the fees.